python/lsst/obs/base/formatters/fitsExposure.py

```diff
@@ -418,21 +422,94 @@ def checkUnrecognized(entry, allowed, description):
         return validated
 
 
-class FitsImageFormatter(FitsExposureFormatter):
+class FitsExposureFormatter(FitsImageFormatterBase):
+    """Specialization for `~lsst.afw.image.Exposure` reading.
+    """
+
+    _readerClass = ExposureFitsReader
+
+    def _compareFilters(self, file_filter_label):
+        """Compare the filter label read from the file with the one in the
+        data ID.
+
+        Parameters
+        ----------
+        file_filter_label : `lsst.afw.image.FilterLabel`, optional
+            Filter label read from the file, if there was one.
+
+        Returns
+        -------
+        filter_label : `lsst.afw.image.FilterLabel` or `None`
+            The preferred filter label; may be the given one or one built from
+            the data ID.
+        """
+        missing = []
+        band = None
+        physical_filter = None
+        if "band" in self.dataId.graph.dimensions.names:
+            band = self.dataId.get("band")
+            if band is None and not self.dataId.hasFull() and "band" in self.dataId.graph.implied.names:
+                missing.append("band")
+        if "physical_filter" in self.dataId.graph.dimensions.names:
+            physical_filter = self.dataId.get("physical_filter")
+            if (physical_filter is None and not self.dataId.hasFull()
+                    and "physical_filter" in self.dataId.graph.implied.names):
+                missing.append("physical_filter")
+        version = self._reader.readSerializationVersion()
+        if missing:
+            # Data ID identifies a filter but the actual filter label values
+            # haven't been fetched from the database; we have no choice but
+            # to use the one in the file.
+            # Warn if that's more likely than not to be bad, because the file
+            # predates filter standardization.
+            if version < 2:
+                warnings.warn(f"Data ID {self.dataId} is missing (implied) value(s) for {missing}; "
```

Comment:
That's a good question, and I simply don't know the answer to which should be preferred.  I feel like `warnings` is what the broader Python world is likely to use, and I know how to write tests against it, so that's what I've used here, but I hadn't noticed the use of `lsst.log` in `FitsRawFormatter`.  Seems like something @timj might have thoughts on.

Reply:
I'm fine with this being logger rather than warnings. I obviously have a preference for `logging` rather than `lsst.log` in butler code. `pipetask` and `cmdLineTask` have to deal with both anyhow.  You can use `assertLogs` in tests if you are using python `logging`.

Reply:
If `warnings` isn't a problem, I'd actually prefer to go with that, here, if only because

> You can use `assertLogs` in tests if you are using python `logging`.

probably evaluates to `False` here in obs_base, and I haven't see a good argument for using `lsst.log` just to warn.

Reply:
Where would those `warnings.warn` messages end up in a production environment? Are they obvious in the log output?

Also you can use `assertLogs` with `lsst.log` too, you just have to wrap it in a `with lsst.log.UsePythonLogging():`.

Reply:
You can wrap but I'd use python logging here anyhow if I was going to use logging. Python warnings are a bit odd in general since people seem to ignore them (see all the tests that warn).

Reply:
I've raised this on community; the dev-guide should probably say something to preempt future versions of this conversation.

I don't have strong opinions, but the community thread doesn't resolve decisively soon, I'm going to merge it as is.